sss/puma.py

```diff
@@ -0,0 +1,232 @@
+    county = Column('county', String)
+    puma_area = Column('puma_area', String)
+    place = Column('place', String, primary_key = True)
+    population_self = Column('population', Integer)
+    weight = Column('weight', Float)
+    year = Column('year', Integer)
+
+
+def read_puma(path, year):
+    """
+    Reads puma data into data frame. The puma data stores by states in txt format.
+    
+    Parameters
+    ----------
+    path: str
+        path name of txt puma file
+    year: int
+    year of the puma data collected
+    
+    Returns
+    -------
+    pandas.datafranme
+        the returned dataframe has different levels of fips, population size etc. from txt file
+    """
+
+    # Please refer 2010_PUMA_Equivalency_Format_Layout to check he record layouts for each field 
+    colspecs = [(0,3),(3,5),(5,13),(13,18),(18,21),(21,29),(29,34),(34,42),(42,47),(47,55),(55,61),(61,70),(70,79),(79,178)]
+    names = ['sl','state_fips','state_national_standards','puma_code','county_fips','county_national_standards',
+        'county_sub_fips','county_sub_national_standards','place_fips','place_national_standards','tract_code',
+        'population','house_number','area_name']
+    # Read a table of fixed-width formatted lines from .txt into DataFrame.
+    df = pd.read_fwf(path,colspecs=colspecs,names=names,dtype=str,encoding='latin1')
+    # change data type of some variables
+    df['population'] = df['population'].astype('float')
+    df['house_number'] = df['house_number'].astype('float')
+    # add a column called year
+    df['year'] = year
+    state_names = {'53': 'WA','10': 'DE','11': 'DC','55': 'WI','54': 'WV','15': 'HI',
+               '12': 'FL','56': 'WY','72': 'PR','34': 'NJ','35': 'NM','48': 'TX',
+               '22': 'LA','37': 'NC','38': 'ND','31': 'NE','47': 'TN','36': 'NY',
+               '42': 'PA','02': 'AK','32': 'NV','33': 'NH','51': 'VA','08': 'CO',
+               '06': 'CA','01': 'AL','05': 'AR','50': 'VT','17': 'IL','13': 'GA',
+               '18': 'IN','19': 'IA','25': 'MA','04': 'AZ','16': 'ID','09': 'CT',
+               '23': 'ME','24': 'MD','40': 'OK','39': 'OH','49': 'UT','29': 'MO',
+               '27': 'MN','26': 'MI','44': 'RI','20': 'KS','30': 'MT','28': 'MS',
+               '45': 'SC','21': 'KY','41': 'OR','46': 'SD'}
+    df['state'] = df['state_fips'].map(state_names)
+    return df
+
+def puma_crosswalk(path, nyc_wa_path, year):
+    """
+    This function is to create a puma crosswalk with county and subcounty.
+    If the puma is in Aew England Area, use sub county(level 797) as sss_place(place).
+    Otherwise the county is assigned to sss_place(place).
+    For State of Washington and New York City, some puma code area names are replace by more meaningful names.
+    
+    Parameters
+    ----------
+    path: str
+        path name of txt puma file
+    nyc_wa_path: str
+        excel file path of nyc and statement of washington replacement name list 
```

Comment:
As written, this file is required even when it's not processing WA or NY data. Could we make this an optional parameter (by setting it equal to a default value like `None` in the function signature) and error if it's needed and not provided?